src/lib/journey/_utilities/metadata.utilities.ts

```diff
@@ -40,17 +46,39 @@ export function buildCallbackMetadata(
       isSelfSubmitting: isSelfSubmitting(callback),
       isUserInputRequired: requiresUserInput(callback),
       idx,
+      social: buildCallbackSocialMetadata(callback),
```

Comment:
Slightly unsure about whether this is the best way to achieve this. The problem I am trying to solve is how to let the stage know that despite this being a self-submitting callback, we still need a "Next" button if the `Select IDP Node` has the `Local Authentication` option enabled.

Reply:
Okay, I think I have a suggestion for handling this in a generic way. I've made some notes locally, but it's getting a bit late for me today, so I'll send them your way tomorrow morning.

The reason this is a unique case because this specific callback is dynamic, and the form structure is not just driven by the existence of the callback, but conditional data inside the callback. We need to be able to handle three types of use-cases:

1. Social IdPs and username-password form
2. Social IdPs and local auth button (no form)
3. Social IdPs only

I'm thinking a prop on the step metadata object of `isUserInputOptional`, which would help inform the stage component to know if it should render the "Next" button or not. This would satisfy the condition of both the step being "self submittable" and having user input that's optional.

To determine this, I think we should have custom functions that can be ran for determining this, but let me play around with this idea tomorrow.

Reply:
Okay, here's what I've got:

1. At the step level, we can iterate over the callback metadata, and add a property of `canForceUserInputOptionality`, or something, which has a boolean value, essentially saying this callback can force normally required user input callbacks to be optional
2. We can then add a new that step property, `isUserInputOptional`, which indicates that the step does indeed have user input type callbacks, but they are optional in this instance

Regarding the first item, I think we should provide a way to analyze callbacks by iterating through the step's callbacks and applying a generic function that could come from a user of the framework, so they could alter this behavior.

```js
// This eventually will be overridable by user of framework
const forceUserInputOptionalityCallbacks = {
  SelectIdPCallback: (callback) => {
    return !!callback
      .getProviders()
      .find((provider) => provider.provider === 'localAuthentication');
  },
};

// Our existing callback metadata builder with extra property to inform
// if any callbacks can override the presence of user input requirements
export function buildCallbackMetadata(
  step: WidgetStep,
  checkValidation: (callback: FRCallback) => boolean,
) {
  return step?.callbacks.map((callback, idx) => {
    return {
      // This one is tough to name, but it really just communicates the
      // ability of a special callback that can self-submit regardless
      // of the presence of user input callbacks
      canForceUserInputOptionality: canForceUserInfoOptionality(callback),
      isFirstInvalidInput: checkValidation(callback),
      isReadyForSubmission: isCbReadyByDefault(callback),
      isSelfSubmitting: isSelfSubmitting(callback),
      isUserInputRequired: requiresUserInput(callback),
      idx,
    };
  });
}

// Out existing step metadata builder with an extra property to inform
// if the step has a unique combination of self-submitting callback as
// well as user input type callbacks
export function buildStepMetadata(callbackMetadataArray) {
  return {
    isStepSelfSubmittable: isStepSelfSubmittable(callbackMetadataArray),
    isUserInputOptional: isUserInputOptional(callbackMetadataArray),
    numOfCallbacks: callbackMetadataArray.length,
    numOfSelfSubmittableCbs: callbackMetadataArray.filter((cb) => !!cb.isSelfSubmitting).length,
    numOfUserInputCbs: callbackMetadataArray.filter((cb) => !!cb.isUserInputRequired).length,
  };
}

export function forceUserInputOptionalityCallbacks(callback) {
  // See if a callback function exists within this collection
  const fn = forceUserInputOptionalityCbs[callback.getType()];

  // If there is a function, run it and it will return a boolean
  return fn && fn();
}

// Notice this function can take a user provided argument function to 
// override behavior (this doesn't have to be well defined)
export function isUserInputOptional(callbackMetadataArray, fn) {
  // default reducer function to check if both overriding callback exists
  // along with user input required callbacks
  const fallbackFn = (prev, curr) => {
    if (curr.canForceUserInfoOptionality && callbackMetadataArray.numOfUserInputCbs > 0) {
      prev = true;
    }
    return prev;
  };
  // Call reduce function with either fallback or user provided function
  return callbackMetadataArray.reduce(fn || fallbackFn, false);
}
```

Now that we have the above information, we can use this to generically affect the rendering of the form. So, if `stepMetadata.isUserInputOptional` is true, then we know we have to render the form. If it's false, then we make sure the step isn't self-submittable. If it isn't the we render the "Next" button.

```html
{#if stepMetadata.isUserInputOptional || !stepMetadata.isStepSelfSubmittable}
  <Button busy={loading} style="primary" type="submit" width="full">
    <T key="nextButton" />
  </Button>
{/if}
```

I hope this all makes sense. Also, please keep in mind the above is just pseudocode and not verified to work.

Reply:
Ok this now makes sense to me - I have a working version reworked to your ideas and I like it. I need to look through what I wrote with a fresh mind but we're most of the way there I think. Thanks for putting so much thought and detail into your response! I'll push it probably tomorrow.